### PR TITLE
Tally Mean Operation

### PR DESCRIPTION
**Pull request by wbinventor**
This short PR adds a new `Tally.average(...)` aggregation operation as a corollary to the `Tally.summation(...)` refactored in #561. This feature is demonstrated in [this](http://nbviewer.jupyter.org/gist/wbinventor/4b25c59e2ed96373bb60) IPython Notebook.

### CODE REVIEW

**Comment by paulromano** [comment]
This looks fine. I will say that `Tally.average()` looks almost exactly like `Tally.summation()` which makes me think a lot of the code ought to be combined somehow. Any ideas how we can do that?

**Comment by wbinventor** [comment]
I agree, I will think a little more about how to best generalize the tally aggregation operations. This may sit on the back-burner for a bit since I need to put in an `Tally.merge(...)` before I can really make use of `Tally.mean(...)` in my data pipeline. Perhaps with that in place I'll have a better idea of how to consolidate things a bit. You can close this PR for now if you'd prefer to wait on that.

**Comment by paulromano** [comment]
No problem keeping it open. Would you put the merge stuff in this PR or another one?

**Comment by wbinventor** [comment]
Well I haven't even started thinking about how to implement the `merge` operation yet, I just know that I need it. It will be at least a week down the road for me; it probably makes sense to include it in this PR since it falls under the general umbrella of tally aggregation.

**Comment by wbinventor** [comment]
This PR has been extensively updated with a lot of new features and bug fixes and I think it's about time to call it a success and open it up for full review. Perhaps @samuelshaner (at least) may review this since most of the code is related to tally operations in the Python API - tally summation, averaging, slicing and merging - which he is very familiar with.

There are three Jupyter Notebooks which illustrate the extensive tally summation, averaging, slicing and merging features that dominate the changes in this PR. The notebooks leverage the H-M benchmark available in the [mit-crpg/benchmarks](https://github.com/mit-crpg/benchmarks/tree/master/mc-performance/openmc) repository. Each of these notebooks is available as a public gist via the following links:

* [Tally Summation](https://gist.github.com/wbinventor/e94925cc7c0b449bcfc4)
* [Tally Averaging](https://gist.github.com/wbinventor/d877e4f1e2ecbeeb7ebb)
* [Tally Slicing / Mering](https://gist.github.com/wbinventor/1f33e41b23ec77870b82)

In summary, the changes introduced here include the following

* **Fast Tally Averaging**
The `Tally.average(...)` method which can be used to average cross one or more filters' bins and/or nuclides and/or scores. I'm not sure how useful this may be to the average user - in my case, it is very useful for computing averages over clustered "distribcell" tallies. As @paulromano pointed out when I opened this PR for the first time, there is a lot of similar looking code between `Tally.summation(...)` and `Tally.average(...)`. This has not improved since most of my recent changes have not been directly related to `Tally.average(...)` but rather the features discussed below. Looking at the two routines, I would opt to keep them separate and somewhat redundant for now since it is not obviously clear how (or whether it is truly valuable) to generalize the two at the moment.

* **Tally Merging**
The Python API's `Tally` class already includes support for "merging" `Tallies` for input generation as a way to improve runtime performance of the code on models with lots of tallies. This PR extends the `Tally.merge(...)` method so that tallies with data - *i.e.*, loaded from statepoint files - can also be merged in downstream data processing to appropriately concatenate the NumPy arrays of data.

* **MGXS Slicing / Merging**
The `openmc.mgxs.MGXS` class now supports `MGXS.merge(...)` and `MGXS.slice(...)` methods which mirror those in the `Tally` class. Again, this may not be generally useful, but I've found it very helpful for managing my "distribcell" MGXS data pipeline to OpenMOC.

* **Tally Slicing**
The Python API's `Tally` class already includes support for "slicing" `Tallies` up by filter bins and/or nuclides and/or scores. This functionality is further improved with some bug patches.

* **Test suite**
The test suite now includes a "test_tally_slice_merge" test to test the features described above.

* **Object Retrieval by Name Methods**
This PR introduces a several new methods to the `Geometry` class to retrieve `Material`, `Cell`, `Universe` and `Lattice` objects by string name.

**Comment by paulromano** [comment]
@wbinventor Looks good to me after a cursory look. It would be great if @samuelshaner is willing to review as well.

I did notice on one of the notebooks that you had 'sum(distribcell)' as a column heading in a dataframe. Should that be just 'distribcell'?

**Comment by wbinventor** [comment]
Which notebook and code cell?

**Comment by wbinventor** [comment]
The columns should be titled "sum(distribcell)" for many of the cases (cases which are summed across distribcell instances), but perhaps something showed up that shouldn't have been summed across.

**Comment by samuelshaner** [comment]
I'm happy to review as well - I'll do a review of this PR this evening.

**Comment by paulromano** [comment]
Tally Summation Out [13]. I was expecting to see "sum" in the data itself rather than the heading as is done for nuclides and scores. Is there a reason to have distribcells different in that respect?

**Comment by wbinventor** [comment]
Oh yes, that is a change I made in this PR. I removed the "sum" and parentheses from the data in each column for filter bin sums since it is redundant and takes space away from displaying the bins summed across. Plus...if I recall correctly, the more important reason is that the `AggregateFilter` now stores the bins in a way that makes it simpler to compare and merge `AggregateFilters` which was not the case before. I can remove the "sum" and parentheses from `AggregateScore` and `AggregateNuclide` as well if need be (cross my fingers).

**Comment by paulromano** [comment]
Got it. Yeah, it would be nice to be consistent whatever you do.

**Comment by wbinventor** [comment]
Agreed. I'll make the columns consistent, along with anything else that
comes up with @samuelshaner's review.

File: openmc/arithmetic.py [hunk 1]
```diff
@@ -868,14 +886,85 @@ def get_pandas_dataframe(self, datasize, summary=None):
...
+
+        # None of the bins in this filter should match in the other filter
+        for bin in self.bins:
+            if bin in other.bins:
+                return False
+
+        # None of the bins in the other filter should match in this filter
+        for bin in other.bins:
+            if bin in self.bins:
+                return False
```
Comment (samuelshaner) [inline openmc/arithmetic.py]: Is this second check redundant? It seems like matching bins would be guaranteed to be found in the check above.

File: openmc/filter.py [hunk 2]
```diff
@@ -77,6 +77,23 @@ def __eq__(self, other):
...
+    def __gt__(self, other):
+        if self.type != other.type:
+            if self.type in _FILTER_TYPES and other.type in _FILTER_TYPES:
+                delta = _FILTER_TYPES.index(self.type) - \
+                        _FILTER_TYPES.index(other.type)
+                return True if delta > 0 else False
+            else:
+                return False
+        else:
+            if 'energy' in self.type and 'energy' in other.type:
```
Comment (samuelshaner) [inline openmc/filter.py]: Could you add a comment that describes the logic of comparing whether one energy filter is greater than another?

File: openmc/filter.py [hunk 3]
```diff
@@ -77,6 +77,23 @@ def __eq__(self, other):
...
+                return self.bins[0] >= other.bins[-1]
+            else:
+                return max(self.bins) > max(other.bins)
+
+    def __lt__(self, other):
+        return not self > other
```
Comment (samuelshaner) [inline openmc/filter.py]: What about equal to?

File: openmc/geometry.py [hunk 4]
```diff
@@ -184,6 +212,227 @@ def get_all_material_universes(self):
...
+            Whether the names must match completely (default is True)
+
+        Returns
+        -------
+        list of openmc.material.Material
+            Materials matching the queried name
+
+        """
+
+        if case_sensitive:
```
Comment (samuelshaner) [inline openmc/geometry.py]: Should this be `if not case_sensitive:`?

File: openmc/tallies.py [hunk 5]
```diff
@@ -2914,6 +3158,157 @@ def summation(self, scores=[], filter_type=None,
...
+
+            # Average across the bins in the user-specified filter
+            for i, self_filter in enumerate(self.filters):
+                if self_filter.type == filter_type:
+                    mean = np.take(mean, indices=bin_indices, axis=i)
+                    std_dev = np.take(std_dev, indices=bin_indices, axis=i)
+                    mean = np.mean(mean, axis=i, keepdims=True)
+                    std_dev = np.mean(std_dev**2, axis=i, keepdims=True)
+                    std_dev /= len(bin_indices)
+                    std_dev = np.sqrt(std_dev)
```
Comment (samuelshaner) [inline openmc/tallies.py]: Should these last two lines be switched?

File: openmc/mgxs/mgxs.py [hunk 6]
```diff
@@ -857,14 +882,172 @@ def get_subdomain_avg_xs(self, subdomains='all'):
...
+        merged_mgxs = copy.deepcopy(self)
+        merged_mgxs._derived = True
+
+        # Merge energy groups
+        if self.energy_groups != other.energy_groups:
+            merged_groups = self.energy_groups.merge(other.energy_groups)
+            merged_mgxs.energy_groups = merged_groups
+
+        # Merge nuclides
+        if self.nuclides != other.nuclides:
```
Comment (samuelshaner) [inline openmc/mgxs/mgxs.py]: In order to merge MGXS, it looks like the MGXS objects either must have the same set of nuclides or the sets of nuclide sets must be mutually exclusive. Is that correct?

File: openmc/tallies.py [hunk 7]
```diff
@@ -760,23 +881,127 @@ def merge(self, tally):
...
         merged_tally.id = None
 
-        # Merge filters
-        for i, filter1 in enumerate(merged_tally.filters):
-            for filter2 in tally.filters:
-                if filter1 != filter2 and filter1.can_merge(filter2):
-                    merged_filter = filter1.merge(filter2)
-                    merged_tally.filters[i] = merged_filter
-                    break
+        # If the two tallies are equal, simpy return copy
```
Comment (samuelshaner) [inline openmc/tallies.py]: simpy -> simply

File: openmc/tallies.py [hunk 8]
```diff
@@ -760,23 +881,127 @@ def merge(self, tally):
...
-        # Add triggers from second tally to merged tally
-        for trigger in tally.triggers:
+        # Add triggers from other tally to merged tally
+        for trigger in other.triggers:
             merged_tally.add_trigger(trigger)
 
+        # If results have not been read, then return tally for input generation
+        if self._results_read is None:
+            return merged_tally
+        #Otherwise, this is a derived tally which needs merged results arrays
```
Comment (samuelshaner) [inline openmc/tallies.py]: `#Otherwise` -> `# Otherwise`

File: openmc/arithmetic.py (hunk 1 above)
Comment (wbinventor) [inline openmc/arithmetic.py]: Good call, I don't think this is needed after all.

File: openmc/filter.py (hunk 2 above)
Comment (wbinventor) [inline openmc/filter.py]: Sure - the short of it is this is needed when merging filters with energy filters. In particular, the order of the tally bin array concatenation is dependent on the result of this comparison.

File: openmc/filter.py (hunk 3 above)
Comment (wbinventor) [inline openmc/filter.py]: What do you mean? The built-in `__le__` method should make use of the customized `__lt__` and `__eq__` methods.

File: openmc/geometry.py (hunk 4 above)
Comment (wbinventor) [inline openmc/geometry.py]: oh wow, yes of course!

Comment (wbinventor) [inline openmc/geometry.py]: fixed this issue here as well as elsewhere in `Geometry`

File: openmc/mgxs/mgxs.py (hunk 6 above)
Comment (wbinventor) [inline openmc/mgxs/mgxs.py]: That is right, that is a limitation of the current implementation. I'll elaborate on this subtle point in the docstring.

File: openmc/tallies.py (hunk 7 above)
Comment (wbinventor) [inline openmc/tallies.py]: got it

File: openmc/tallies.py (hunk 8 above)
Comment (wbinventor) [inline openmc/tallies.py]: got it

File: openmc/tallies.py (hunk 5 above)
Comment (wbinventor) [inline openmc/tallies.py]: I don't think so - my implementation here reflects my understanding of the equation for the standard deviation in the table of "Example Formulas" in cell (3,2) on this [wiki](https://en.wikipedia.org/wiki/Propagation_of_uncertainty) page. The first line calls `np.mean(...)` of the squared standard deviation (variance) divided by n, the second line applies the second (1/n) factor to get the squared coefficient in the equation, and the final line takes the square root of the sum to back out the final standard deviation. If I

**Comment by wbinventor** [comment]
I've updated this PR per your comments @samuelshaner and @paulromano.

**Pull request closed without merging**